clang/lib/AST/ASTImporter.cpp

```diff
@@ -3018,7 +3032,7 @@ ExpectedDecl ASTNodeImporter::VisitRecordDecl(RecordDecl *D) {
         for (auto *R : Redecls) {
           auto *RI = cast<CXXRecordDecl>(R);
           if (R != Redecls.front() ||
-              !isa<InjectedClassNameType>(RI->getTypeForDecl()))
+              !isa_and_nonnull<InjectedClassNameType>(RI->getTypeForDecl()))
             RI->setTypeForDecl(nullptr);
           // This function tries to get the injected type from getTypeForDecl,
```

Comment:
A better solution is to find the first `InjectedClassNameType` in the chain (if there is one), and use this type for all objects in the chain (that had previously `nullptr` or non-`InjectedClassNameType` `TypeForDecl`).

Reply:
> If there was some AST reference to the existing old TypeForDecl object this becomes invalid and may cause bugs later, probably only in specific rare situations.

Thanks for pointing that out, I agree with that analysis

> A better solution is to find the first InjectedClassNameType in the chain (if there is one), and use this type for all objects in the chain (that had previously nullptr or non-InjectedClassNameType TypeForDecl).

Right, you're suggesting dropping the `R != Redecls.front()` requirement here and instead change it to `R != DeclWithValidTypeForDecl`. I'll try that out

Reply:
We may have a redeclaration chain like this (TypeForDecl is: N=nullptr, I=`InjectedClassNameType`, T=not `InjectedClassNameType`):
N->N->I->T->T
(I do not know what happens practically, probably there is only a N->N->T like case, and I->T->T like cases.)
We should find the "I" object in this chain and put this `InjectedClassNameType` object into all others. If there is not an "I" object, `getTypeDeclType` can be called for the first, this creates it.

Reply:
To clarify issue, given a chain `N1 <- N2 <- I <- T1 <- T2` (where `N1` is the front of the chain) using the current implementation would be:
```
// Iteration 0
setTypeForDecl(N1, nullptr)
getInjectedClassNameType(N1,...) creates new InjectedClassNameType on N1

// Iteration 1
setTypeForDecl(N2, nullptr)
getInjectedClassNameType(N2,...) creates new InjectedClassNameType on N2

// Iteration 2
getInjectedClassNameType(I,...) sets I->TypeForDecl=T1->TypeForDecl

// Iteration 3
setTypeForDecl(T1, nullptr)
getInjectedClassNameType(T2,...) creates new InjectedClassNameType on T1

// Iteration 4
setTypeForDecl(T2, nullptr)
getInjectedClassNameType(T2,...) creates new InjectedClassNameType on T2
```

Which leaves `I` with an incorrect `TypeForDecl`. Is this the scenario you were concerned about?

It seems like what we want is something like:
```
for (auto * Redecl : Redecls) {
  // If we have an InjectedClassNameType on the chain already, copy it to all other decls.
  if (FirstInjectedTypeInChain) {
    Redecl->TypeForDecl = FirstInjectedTypeInChain->TypeForDecl;
  } else {
  // Otherwise, create a new InjectedClassNameType and propagate it to the rest of the chain
  ...
  }
}
```

Though I don't think we can implement the `Otherwise` case with just calling `getTypeDeclType`, because it only creates non-`InjectedClassNameType`s (unless we already have a redeclaration with a `InjectedClassNameType`)

Reply:
This was a mistake, I meant `getInjectedClassNameType` (and not `getTypeDeclType`) in the previous comment.
The `getInjectedClassNameType` does not create a new type if the previous decl (`getPreviousDecl`) has a `TypeForDecl`. If we start from the first decl only for the first a new will be created, for the others it is copied from the previous decl. But the function does not expect a non-`InjectedClassNameType`, in this case we get a crash unless the typeForDecl is set to nullptr before.
One thing to ensure is that all declarations have the same TypeForDecl object. Other thing is that if there is an existing `InjectedClassNameType` in any redeclaration, this should be reused and not overwritten with a new one. So in this second case it should be possible to get the existing `InjectedClassNameType` and set it with `setTypeForDecl` in the others.

Reply:
Tried fixing this in the latest iteration